scrape/scrape.go

```diff
@@ -1527,6 +1552,16 @@ loop:
 				targetScrapePoolExceededLabelLimits.Inc()
 				break loop
 			}
+
+			sl.cache.metaMtx.Lock()
```

Comment:
Hmmm, those locks are a bit scary here. I assume lock is here because we want the latest metadata across all sources of a single metric name, right?  Which could be inconsistent during scrape. Or am I wrong?

Nevertheless, we are on the hot path of scraping which has to be efficient, so we have to ideally benchmark this path. WDYT? Ideally with cases of potential lock congestion. If there is regression, we could opt in for some less consistent framework for choosing what metadata should stay.


Also for "error?" cases where process A is registering `__name__=metric1` with help `ABC` and process B registering   `__name__=metric1` with help `ABC2`, we will constantly update those metadata back and worth in worse case, right?

Reply:
Sorry @bwplotka  for the super late reply.

Yes, I think we can go with a feature flag until we see its effects on larger-scale Prometheus deployments.

> I assume lock is here because we want the latest metadata across all sources of a single metric name, right? Which could be inconsistent during scrape. Or am I wrong?

Yes, that's what we hold as metadata in the scrapeCache. This inconsistency will happen when two different scrape_configs share a common metric name and try to simultaneously update its metadata, right? I think that yes, in the worst case that two processes register different metadata in each scrape iteration, we will be recording all these updates.

> so we have to ideally benchmark this path. WDYT? Ideally with cases of potential lock congestion

I'm building a benchmark to measure the effect and will come back to you. Maybe we can also consider a prombench run afterwards just to be certain

Reply:
`if I'm getting this right, the scrapeCache is unique per scrapeLoop.` this is how I read the code as well. The `newLoop` function, part of `scrapePool`, takes in the `scrapeLoopOptions` struct. If the cache field on that struct is nil a new scrape cache is created. When a sync happens and the new scrape loops are started (some of which won't really have changed) the existing scrape loops are passed where/when relevant.